### PR TITLE
Add Spark 3.0 EMR Shim layer 

### PR DESCRIPTION
**Pull request by tgravescs**
closes https://github.com/NVIDIA/spark-rapids/issues/818

This adds a shim layer to support Amazon EMR.
The main issues is they have a different version number and then their FileScanRDD takes a different number of parameters. 

Other then that add some tests support and fix a few things that assumed DataSource v1 was enabled.  There is one join test that has a follow up jira because it appears they have a new exec for bloom filters.

I ran all the integration tests successfully on an EMR cluster by specifying --runtime_env emr with the spark-submit command.  It does not need to be built on an EMR cluster because we are using reflection to determine the FileScanRDD parameter count.

I did not run the unit tests against it as there is another jira to setup build and test on EMR and we can do it with that.

### CODE REVIEW

**Comment by tgravescs** [comment]
build

File: integration_tests/README.md [hunk 1]
```diff
@@ -58,6 +58,11 @@ All of the tests will run in a single application.  They just enable and disable
 
 You do need to have access to a compatible GPU with the needed CUDA drivers. The exact details of how to set this up are beyond the scope of this document, but the Spark feature for scheduling GPUs does make this very simple if you have it configured.
 
+### Runtime Environment
+
+`--runtime_env` is used to specify the environment you are running the tests in. generally Valid values are `databricks` and `emr`. This is generally used
```
Comment (jlowe) [inline integration_tests/README.md:63]:
```suggestion
`--runtime_env` is used to specify the environment you are running the tests in. Valid values are `databricks` and `emr`. This is generally used
```

File: integration_tests/README.md [hunk 2]
```diff
@@ -58,6 +58,11 @@ All of the tests will run in a single application.  They just enable and disable
...
+when certain environment have different behavior.
```
Comment (jlowe) [inline integration_tests/README.md:64]:
```suggestion
when certain environments have different behavior.
```

**Comment by tgravescs** [comment]
build

**Review by jlowe — APPROVED** [review]

File: integration_tests/README.md [hunk 3]
```diff
@@ -58,6 +58,11 @@ All of the tests will run in a single application.  They just enable and disable
 
 You do need to have access to a compatible GPU with the needed CUDA drivers. The exact details of how to set this up are beyond the scope of this document, but the Spark feature for scheduling GPUs does make this very simple if you have it configured.
 
+### Runtime Environment
+
+`--runtime_env` is used to specify the environment you are running the tests in. Valid values are `databricks` and `emr`. This is generally used
+when certain environments have different behavior.
```
Comment (revans2) [inline integration_tests/README.md:64]: nit: Can we detect `emr` at runtime? I know for databricks it is hard to detect so we avoid doing it directly in the tests.

I am fine with doing it on the command line for consistency.

File: integration_tests/src/main/python/parquet_test.py [hunk 4]
```diff
@@ -62,7 +62,8 @@ def test_parquet_fallback(spark_tmp_path, read_func, disable_conf):
     assert_gpu_fallback_collect(
             lambda spark : reader(spark).select(f.col('*'), f.col('_c2') + f.col('_c3')),
             'FileSourceScanExec',
-            conf={disable_conf: 'false'})
+            conf={disable_conf: 'false',
+                "spark.sql.sources.useV1SourceList": "parquet"})
```
Comment (revans2) [inline integration_tests/src/main/python/parquet_test.py:66]: Do we need/want follow on issues for datasource V2 vs V1 testing?

File: shims/spark300emr/src/main/scala/com/nvidia/spark/rapids/shims/spark300emr/Spark300EMRShims.scala [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      sparkSession: SparkSession,
+      readFunction: (PartitionedFile) => Iterator[InternalRow],
+      filePartitions: Seq[FilePartition]): RDD[InternalRow] = {
+
+    val tclass = classOf[org.apache.spark.sql.execution.datasources.FileScanRDD]
+    val constructors = tclass.getConstructors()
+    if (constructors.size > 1) {
+      throw new IllegalStateException(s"Only expected 1 constructor for FileScanRDD")
+    }
+    val constructor = constructors(0)
```
Comment (revans2) [inline shims/spark300emr/src/main/scala/com/nvidia/spark/rapids/shims/spark300emr/Spark300EMRShims.scala:47]: Do we care about caching the constructor?  I assume it is not a big performance hit right now so it is probably not a big deal.

**Review by revans2 — COMMENTED** [review]
Looks good, just a few questions.

File: integration_tests/README.md (hunk 3 above)
Comment (tgravescs) [reply integration_tests/README.md:64]: yeah I did it for consistency and if you noticed here we actually have another check for 3.0.0. I was going to file a follow on try to commonize as I think we could use the shim layer for it all now.

File: integration_tests/src/main/python/parquet_test.py (hunk 4 above)
Comment (tgravescs) [reply integration_tests/src/main/python/parquet_test.py:66]: yeah I'll file one

**Review by revans2 — DISMISSED** [review]

File: shims/spark300emr/src/main/scala/com/nvidia/spark/rapids/shims/spark300emr/Spark300EMRShims.scala (hunk 5 above)
Comment (tgravescs) [reply shims/spark300emr/src/main/scala/com/nvidia/spark/rapids/shims/spark300emr/Spark300EMRShims.scala:47]: I honestly didn't think about it as I didn't expect it to be called real often, but its definitely not a bad idea. I'll take a quick measurement

File: integration_tests/README.md (hunk 3 above)
Comment (tgravescs) [reply integration_tests/README.md:64]: https://github.com/NVIDIA/spark-rapids/issues/830

File: integration_tests/src/main/python/parquet_test.py (hunk 4 above)
Comment (tgravescs) [reply integration_tests/src/main/python/parquet_test.py:66]: https://github.com/NVIDIA/spark-rapids/issues/829

File: shims/spark300emr/src/main/scala/com/nvidia/spark/rapids/shims/spark300emr/Spark300EMRShims.scala (hunk 5 above)
Comment (tgravescs) [reply shims/spark300emr/src/main/scala/com/nvidia/spark/rapids/shims/spark300emr/Spark300EMRShims.scala]: updated to cache. very minimal but better perf and I coded it so might as well.

**Comment by tgravescs** [comment]
build

**Review by jlowe — APPROVED** [review]

**Review by revans2 — APPROVED** [review]

**Pull request merged**